### ISSUE TITLE
hard mode have HUGE bug when spy is stabbed,there are some ninjas don't move anymore

### ISSUE THREAD

**Issue by enochhz** [Jester565/SpyNinjaGame#34]
# Situation 1: 2 ninjas around briefcase room don't move anymore

<img width="402" alt="screen shot 2016-11-23 at 11 50 56 am" src="https://cloud.githubusercontent.com/assets/17519068/20576587/573534c8-b173-11e6-9a70-2eec6af80b58.png">
<img width="415" alt="screen shot 2016-11-23 at 11 51 13 am" src="https://cloud.githubusercontent.com/assets/17519068/20576590/5847c70e-b173-11e6-85fb-c209d627ebb2.png">
<img width="393" alt="screen shot 2016-11-23 at 11 51 30 am" src="https://cloud.githubusercontent.com/assets/17519068/20576592/59b30a36-b173-11e6-9437-77cd034fd051.png">

**Comment by enochhz** [comment]
Also AI mode is acting wired, ninja can only check four direction and move to the direction if there is a spy on the direction @jtara1

**Comment by enochhz** [comment]
And I realize you replaced the AI method I push yesterday. There is the same situation happened in my AI design yesterday, and I fixed it, it is nonpointer problem in my stituation. @jtara1 
Have you check my AI yesterday?

**Comment by enochhz** [comment]
# Situation 2: ninja position: x:7 y: 3 don't move in hardmode

<img width="433" alt="screen shot 2016-11-23 at 11 44 05 am" src="https://cloud.githubusercontent.com/assets/17519068/20577182/02a6fb50-b176-11e6-92c1-b92ccbb14bf1.png">

**Comment by enochhz** [comment]
# Situation 3: Exception
<img width="629" alt="screen shot 2016-11-23 at 12 17 28 pm" src="https://cloud.githubusercontent.com/assets/17519068/20577375/d306cbae-b176-11e6-805f-d24a23629f76.png">

**Comment by jtara1** [comment]
I believe the ninja stuck above the Room on the middle-right side was because I needed to set the `destinationCoordinate` of the ninja to null.

Commit 629f457a6d2afc9bb36266497191754b1c279638 should solve **situation 1** and **situation 2**

**Comment by jtara1** [comment]
As for **situation 3** please `git pull` to make sure you have the latest commits I've pushed.

More specifically check the Ninja class to see if that method is there that the java.lang.Error mentions.

**Comment by enochhz** [comment]
# Situation 4: Spy don't lose lives in hard mode

<img width="456" alt="screen shot 2016-11-23 at 12 26 48 pm" src="https://cloud.githubusercontent.com/assets/17519068/20577654/31b3f2de-b178-11e6-9ba8-f9f1288fe994.png">

**Comment by enochhz** [comment]
????
I add a simple AI enemyLook() method in GameEngine yesterday, and every thing works fine, have you check those commits? @jtara1

**Comment by jtara1** [comment]
Yea give me a minute, I'm going to create a new branch going back to your implementation of the ninja ai to compare the two implementations.

There may be a few different bugs at that point that I fixed today. I probably introduced some bugs myself with my changes too. I'm in the process of identifying and fixing though.

**Comment by jtara1** [comment]
Update: Here's the branch from yesterday before I merged and rewrote the ninja ai 

https://github.com/Jester565/SpyNinjaGame/tree/hao-ninja-ai

**Comment by jtara1** [comment]
**Situation 4 Spy doesn't lose lives in hard mode**

What are the steps to reproduce the issue?

I'm in hard & debug mode like you, but it works. Again, I probably fixed this in a recent commit today (11/23/16).

```
Move Status: Moved!
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Lives: 1
[ ][R][ ][ ][R][ ][N][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][B][N][ ]   Invincible Turns: 0
[ ][ ][N][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][ ][ ][b][ ][ ][R][ ]   ---
[ ][ ][S][ ][ ][ ][ ][ ][ ]   Hard Mode: Enabled
[ ][ ][N][ ][ ][ ][r][ ][ ]   Debug Mode: Enabled
[ ][R][N][I][R][ ][ ][R][ ]
[N][ ][ ][ ][ ][ ][ ][ ][ ]

Enter a direction to move or another command
W: Up | A: Left | S: Down | D: Right
1: Shoot | 2: Debug | 3: More Options | 4: HardMode
w
Move Status: Moved!
Game Over
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Lives: 0
[ ][R][ ][ ][R][ ][N][R][ ]   Ammo: 1
[ ][ ][ ][ ][ ][ ][B][N][ ]   Invincible Turns: 0
[ ][ ][N][ ][ ][ ][ ][ ][ ]   Radar: Disabled
[ ][R][S][ ][b][ ][ ][R][ ]   ---
[ ][ ][ ][ ][ ][ ][ ][ ][ ]   Hard Mode: Enabled
[ ][ ][N][ ][ ][ ][r][ ][ ]   Debug Mode: Enabled
[ ][R][N][I][R][ ][ ][R][ ]
[N][ ][ ][ ][ ][ ][ ][ ][ ]

Select an option:
1: Start New Game
2: Load Game
3: Exit
```

**Comment by enochhz** [comment]
this should help you for figure out my AI designing 
The only two thing i did for AI is:

1. f17d9160637a61025b6e1bbea4493ff5d99ed028
check If the game is hardmode: yes: enemy look all direction and move smart 
                                                      no: move randomly(original code already have that)
2. 84485c89ba051bd6f4d4d158cc4785dad2525320
enemyLook() method: check if there is any spy on the direction. This method can only be accessed when hardmode is true.

**Comment by jtara1** [comment]
Both of our AI's have

* an `enemyLook(...)` which does something similar
* feature that moves ninja towards spy if spotted otherwise moves spy in random direction 

Your AI is more simple and works well.

From the looks of it your AI

* ninja sight (or look range) is as long as the grid
* only moves toward the spy if spotted the same turn

My AI is capable of 

* the ninja sight (look range) and directions it looks in can vary
* tracking the spy's last known whereabouts and moving towards that coordinate until blocked by a `Room` or Grid boundary or it reaches that coordinate.

edit: It might be a good idea to create a new issue discussing how AI should behave.

**Comment by enochhz** [comment]
I copied all files until 82bb96b3761ff60820f1921629e1e89f5318cc7e and running on my own branch, and haven't found any bug yet for the hard mode, there is no bug as I commited in f17d9160637a61025b6e1bbea4493ff5d99ed028 for enemyMove() method as well.
the AI design i did is:  when it is the turn for enemy, the enemy look for spy for 4 directions, move to the direction which there is spy, if there is a room on his direction, the enemy's vision will be block, if the enemy can't find any spy on 4 directions, just move randomly. 

anyway, Many bugs we have now are the bugs i just fixed until 82bb96b3761ff60820f1921629e1e89f5318cc7e,it is ok to have different designing of AI, just make make sure it works.

And sorry about can't work with you on fixing bugs now(I have class until 8pm), the suggestion I give you is making two hardMode in UI 1.HardMode(which I did) 2.HardMode2(which you did)
and text two modes to see which works out better. If mine work fine as I said, you can just change the code back to original AI design, and don't need to spend time on figuring out handling the bug master files are having now.

**Comment by jtara1** [comment]
In my branch https://github.com/Jester565/SpyNinjaGame/commits/james-ninja-ai I've added a 

> Change Difficulty 

feature and the AI will behave slightly differently for each difficulty.

| Difficulty | Dir. look | Sight Range | Behavior |
| --- | --- | --- | --- |
| easy | n/a | n/a | move randomly |
| medium | all 4 | 9 | move toward spy if spotted otherwise randomly (Hao's AI) |
| hard | all 4 | 2 | move toward spy or last known position of spy otherwise move randomly |
| very hard | all 4 | 9 | move toward spy or last known position of spy otherwise move randomly |

* I don't see any bugs in my branch although sometimes a ninja will stop following spy after a while (on hard or very hard) - I'm investigating this. I haven't updated save / load features for the AI so that would be a bug I suppose.

**Comment by jtara1** [comment]
Closing with merge #37
